### PR TITLE
feat: implement dev mode

### PR DESCRIPTION
**Pull request by aeng**
https://zanata.atlassian.net/browse/ZNTAMT-2

---
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-mt/21)

### CODE REVIEW

**Comment by huangp** [comment]
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[pom.xml, line 697 at r1](https://reviewable.io:443/reviews/zanata/zanata-mt/21#-KmTf6dNiib4TfmA4JFd:-KmTf6dNiib4TfmA4JFe:blojnxb) ([raw file](https://github.com/zanata/zanata-mt/blob/2aa317dd6c8d7fd5298f5cc3ddd70140f8d60209/pom.xml#L697)):*
> ```XML
>   <profiles>
>     <profile>
>       <id>production</id>
> ```

It's better to use property to control profile activation. This way it's easier to control multiple profiles.

---

*[README.md, line 14 at r1](https://reviewable.io:443/reviews/zanata/zanata-mt/21#-KmTesOe4fLZGkj1Y_vG:-KmTesOe4fLZGkj1Y_vH:b-44km4a) ([raw file](https://github.com/zanata/zanata-mt/blob/2aa317dd6c8d7fd5298f5cc3ddd70140f8d60209/README.md#L14)):*
> ```Markdown
> ## PROD mode
> PROD mode can be enabled by activating maven profile `production`. This will allow service to use Microsoft Translator service with provided **ZANATA_MT_AZURE_KEY**
> e.g. `mvn clean install -Pproduction`.
> ```

It's better to enable a profile using property instead of profile id. e.g. ```-Dproduction```

---

*[src/main/java/org/zanata/mt/service/PersistentTranslationService.java, line 183 at r1](https://reviewable.io:443/reviews/zanata/zanata-mt/21#-KmTfYQ2KMQEJJxQsbyU:-KmTfYQ2KMQEJJxQsbyV:bhe1myg) ([raw file](https://github.com/zanata/zanata-mt/blob/2aa317dd6c8d7fd5298f5cc3ddd70140f8d60209/src/main/java/org/zanata/mt/service/PersistentTranslationService.java#L183)):*
> ```Java
>                     .translate(sources, fromLocaleCode, toLocaleCode,
>                             mediaType);
>         }
> ```

I think a cleaner way will be to have the mockTranslatorBackend implement the same interface as MicrosoftTranslatorbackend. And the profile will control which one to package into the final war. You can also choose to package different beans.xml with ```@Alternative``` if you want to do the CDI way. That way you don't need a if else statement in code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/21)*

**Comment by huangp** [comment]
Reviewed 12 of 12 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/main/java/org/zanata/mt/api/service/impl/DocumentResourceImpl.java, line 117 at r2](https://reviewable.io:443/reviews/zanata/zanata-mt/21#-KmU6lHtSs-TIKCIV8Ny:-KmU6lHtSs-TIKCIV8Nz:boavdci) ([raw file](https://github.com/zanata/zanata-mt/blob/9ff4dfc79301bdbbebab6e796eb46cc215962719/src/main/java/org/zanata/mt/api/service/impl/DocumentResourceImpl.java#L117)):*
> ```Java
>         if (!zanataMTStartup.isDevMode()) {
>             backendID = BackendID.MS;
>         }
> ```

Here we still need to look up a build time property (unchangeable at runtime) for every single request. My idea was to package only one type of TranslatorBackend in the war since the dev mode flag is decided at build time. So we don't need to check anything at run time unless you want to make this dev mode flag a runtime switch (like what we have in Zanata server config)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/21)*

**Comment by aeng** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/main/java/org/zanata/mt/api/service/impl/DocumentResourceImpl.java, line 117 at r2](https://reviewable.io:443/reviews/zanata/zanata-mt/21#-KmU6lHtSs-TIKCIV8Ny:-KmUMhobCkemmnWvz9sI:b-9ylfw7) ([raw file](https://github.com/zanata/zanata-mt/blob/9ff4dfc79301bdbbebab6e796eb46cc215962719/src/main/java/org/zanata/mt/api/service/impl/DocumentResourceImpl.java#L117)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

Here we still need to look up a build time property (unchangeable at runtime) for every single request. My idea was to package only one type of TranslatorBackend in the war since the dev mode flag is decided at build time. So we don't need to check anything at run time unless you want to make this dev mode flag a runtime switch (like what we have in Zanata server config)
</blockquote></details>

My idea was handle it as another type of backendservice which potentially might be multiple in the future to allow user to select during request. We gonna need runtime check for every request when that comes anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/21)*

**Comment by huangp** [comment]
:lgtm: -(Awaiting review) +Reviewed +(On QA) 

---

Review status: 14 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/main/java/org/zanata/mt/api/service/impl/DocumentResourceImpl.java, line 117 at r2](https://reviewable.io:443/reviews/zanata/zanata-mt/21#-KmU6lHtSs-TIKCIV8Ny:-KmUQHc1XNFfpJJVgjDn:b-biie6n) ([raw file](https://github.com/zanata/zanata-mt/blob/9ff4dfc79301bdbbebab6e796eb46cc215962719/src/main/java/org/zanata/mt/api/service/impl/DocumentResourceImpl.java#L117)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

My idea was handle it as another type of backendservice which potentially might be multiple in the future to allow user to select during request. We gonna need runtime check for every request when that comes anyway.
</blockquote></details>

Alright then

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/21#-:-KmUQMtHJAWT3J2QDmA-:b49ms4)*

**Comment by huangp** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/21)*

**Comment by huangp** [comment]
Reviewed 11 of 11 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[README.md, line 4 at r4](https://reviewable.io:443/reviews/zanata/zanata-mt/21#-KmZSPDx0qRv9OGhiYf1:-KmZSPDx0qRv9OGhiYf2:b-d6cu7t) ([raw file](https://github.com/zanata/zanata-mt/blob/2f67d2c478b527fee35d47613965e5eb32598720/README.md#L4)):*
> ```Markdown
> 
> ## DEV mode
> DEV mode is enabled by default. The service will not use MS backend in this mode but will return string with prefix 'translated'. **ZANATA_MT_AZURE_KEY**
> ```

Is Azure key needed here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/21)*

**Comment by aeng** [comment]
Review status: 19 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[README.md, line 4 at r4](https://reviewable.io:443/reviews/zanata/zanata-mt/21#-KmZSPDx0qRv9OGhiYf1:-KmZTWInoyPLhPrPOez_:boe266g) ([raw file](https://github.com/zanata/zanata-mt/blob/2f67d2c478b527fee35d47613965e5eb32598720/README.md#L4)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

Is Azure key needed here?
</blockquote></details>

good catch, forgot to finish the sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/21)*

**Comment by huangp** [comment]
Reviewed 11 of 11 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-mt/21)*

File: src/main/java/org/zanata/mt/service/ZanataMTStartup.java [hunk 1]
```diff
@@ -27,49 +23,35 @@
...
-            LOG.warn("Cannot load build info");
+        LOG.info("===================================");
+        LOG.info("===================================");
+        LOG.info("=== Machine Translation Service ===");
+        LOG.info("===================================");
+        LOG.info("===================================");
+        LOG.info("Build info: version-" + configurationService.getVersion() +
+                " date-" + configurationService.getBuildDate());
+        if (configurationService.isDevMode()) {
+            LOG.warn("THIS IS DEV MODE BUILT. DO NOT USE IT FOR PRODUCTION");
```
Comment (djansen-redhat) [inline src/main/java/org/zanata/mt/service/ZanataMTStartup.java]: THIS IS **A** DEV MODE BUIL**D**

Comment (aeng) [inline src/main/java/org/zanata/mt/service/ZanataMTStartup.java]: Done.

**Comment by djansen-redhat** [comment]
Verified. Just waiting on @huangp's appropval

**Pull request merged**